frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.tsx

```diff
@@ -139,85 +133,51 @@ function ActionCreatorComponent({
     setQuestion(question.setCard(actionCard));
     setTimeout(() => setShowSaveModal(false), 1000);
     onClose?.();
   };
 
-  const handleClose = () => setShowSaveModal(false);
+  const handleCloseNewActionModal = () => setShowSaveModal(false);
 
-  const handleExampleClick = () => {
+  const handleClickExample = () => {
     setQuestion(
       question.setQuery(query.setQueryText(query.queryText() + EXAMPLE_QUERY)),
     );
   };
 
+  const saveModalElement = showSaveModal ? (
+    <Modal title={t`New Action`} onClose={handleCloseNewActionModal}>
+      <CreateActionForm
+        question={question}
+        formSettings={formSettings as ActionFormSettings}
+        modelId={defaultModelId}
+        onCreate={handleSave}
+        onCancel={handleCloseNewActionModal}
+      />
+    </Modal>
+  ) : null;
+
   return (
-    <>
-      <Modal wide onClose={onClose}>
-        <ModalRoot>
-          <ActionCreatorBodyContainer>
-            <ModalLeft>
-              <DataReferenceTriggerButton onClick={toggleDataRef} />
-              <ActionCreatorHeader
-                type="query"
-                name={question.displayName() ?? t`New Action`}
-                onChangeName={newName =>
-                  setQuestion(q => q.setDisplayName(newName))
-                }
-              />
-              <EditorContainer>
-                <QueryActionEditor
-                  question={question}
-                  setQuestion={setQuestion}
-                />
-              </EditorContainer>
-              <ModalActions>
-                <Button onClick={onClose} borderless>
-                  {t`Cancel`}
-                </Button>
-                <Button
-                  primary
-                  disabled={query.isEmpty()}
-                  onClick={handleSaveClick}
-                >
-                  {isNew ? t`Save` : t`Update`}
-                </Button>
-              </ModalActions>
-            </ModalLeft>
-
-            <DataReferenceInline
-              isOpen={isDataRefOpen}
-              onClose={closeDataRef}
-            />
-
-            {!isDataRefOpen && (
-              <FormCreator
-                params={question?.parameters() ?? []}
-                formSettings={
-                  question?.card()?.visualization_settings as ActionFormSettings
-                }
-                onChange={setFormSettings}
-                onExampleClick={handleExampleClick}
-              />
-            )}
-          </ActionCreatorBodyContainer>
-        </ModalRoot>
-      </Modal>
-      {showSaveModal && (
-        <Modal title={t`New Action`} onClose={handleClose}>
-          <CreateActionForm
-            question={question}
-            formSettings={formSettings as ActionFormSettings}
-            modelId={defaultModelId}
-            onCreate={handleOnSave}
-            onCancel={handleClose}
-          />
-        </Modal>
-      )}
-    </>
+    <ActionCreatorView
+      isNew={isNew}
+      isDataRefOpen={isDataRefOpen}
```

Comment:
I think it would be worth moving the data ref open/close state to `ActionCreatorView` — looks as a view layer concern to me

Reply:
Good call. I think I've made a mess, this PR isn't well structured, since there're some other refactoring in https://github.com/metabase/metabase/pull/27809 as well.

I'll try to address all your comments where application, but I need to move some commits from that branch to this PR first.

Reply:
Done!